src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -3267,10 +3267,18 @@ LogicalResult ONNXQuantizeLinearOp::inferShapes(
   auto yTy = y().getType().cast<ShapedType>();
 
   if (!yTy.hasStaticShape()) {
-    // TODO: Unfortunately, we can't tell if this should be signed or
-    // unsigned here...
-    IntegerType i8Type = IntegerType::get(getContext(), 8);
-    RankedTensorType outType = RankedTensorType::get(inTy.getShape(), i8Type);
+    auto zero = y_zero_point();
+    RankedTensorType outType;
+    // If zero point type isn't provided, output type defaults to ui8.
+    if (isFromNone(zero)) {
+      auto builder = mlir::Builder(this->getContext());
+      IntegerType ui8Type = builder.getIntegerType(8, false);
+      outType = RankedTensorType::get(inTy.getShape(), ui8Type);
+    } else {
+      // If zero point is provided, output type is same as zero point type.
+      auto zpTy = zero.getType().cast<ShapedType>().getElementType();
+      outType = RankedTensorType::get(inTy.getShape(), zpTy);
```

Comment:
nit: to make naming consistent `zpTy` should refer to its tensor type, so you should write:
```
auto zpTy = zero.getType().cast<ShapedType>();
outType = RankedTensorType::get(inTy.getShape(), zpTy.getElementType());
```